http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/Queries.scala

```diff
@@ -0,0 +1,146 @@
+        case (clusterAndLabel, _) =>
+          labelQuery
+            .filter {
+              _.clusterId === clusterAndLabel._1.id
+            }
+            // The following confusing line is equivalent to the much simpler:
+            // .filter { lbl => (lbl.key, lbl.value) inSetBind labelMap.toSet }
+            // Unfortunately slick doesn't support inSet/inSetBind for tuples.
+            // https://github.com/slick/slick/issues/517
+            .filter { lbl =>
+              labelMap.map { case (k, v) => lbl.key === k && lbl.value === v }.reduce(_ || _)
+            }
+            .length === labelMap.size
+      }
+    }
+
+    res.result.map {
+      x =>
+        val clusterLabelMap: Map[(ClusterRecord, Option[RuntimeConfig]), Map[String, Chain[String]]] = x.toList.foldMap {
+          case (clusterRecordAndLabel, runTimeConfig) =>
+            val labelMap = clusterRecordAndLabel._2.map(labelRecord => labelRecord.key -> Chain(labelRecord.value)).toMap
+            Map((clusterRecordAndLabel._1, runTimeConfig.map(_.runtimeConfig)) -> labelMap)
+        }
+
+        clusterLabelMap.map {
+          case (clusterAndRuntimeConfig, labelMp) =>
+            val lmp = labelMp.mapValues(_.toList.toSet.head)
+            val dataprocInfo = (clusterAndRuntimeConfig._1.googleId, clusterAndRuntimeConfig._1.operationName, clusterAndRuntimeConfig._1.stagingBucket).mapN {
+              (googleId, operationName, stagingBucket) =>
+                DataprocInfo(googleId, OperationName(operationName), GcsBucketName(stagingBucket), clusterAndRuntimeConfig._1.hostIp map IP)
+            }
+
+            val serviceAccountInfo = ServiceAccountInfo(
+              clusterAndRuntimeConfig._1.serviceAccountInfo.clusterServiceAccount.map(WorkbenchEmail),
+              clusterAndRuntimeConfig._1.serviceAccountInfo.notebookServiceAccount.map(WorkbenchEmail)
+            )
+            ListClusterResponse(
+              clusterAndRuntimeConfig._1.id,
+              ClusterInternalId(clusterAndRuntimeConfig._1.internalId),
+              clusterAndRuntimeConfig._1.clusterName,
+              clusterAndRuntimeConfig._1.googleProject,
+              serviceAccountInfo,
+              dataprocInfo,
+              clusterAndRuntimeConfig._1.auditInfo,
+              clusterAndRuntimeConfig._2.getOrElse(throw new Exception(s"no runtimeConfig found for ${clusterAndRuntimeConfig._1.id}")), //In theory, the exeception should never happen because it's enforced by db foreign key
+              Cluster.getClusterUrl(clusterAndRuntimeConfig._1.googleProject, clusterAndRuntimeConfig._1.clusterName, Set.empty, lmp), //TODO: remove clusterImages field
+              ClusterStatus.withName(clusterAndRuntimeConfig._1.status),
+              lmp,
+              clusterAndRuntimeConfig._1.jupyterExtensionUri,
+              clusterAndRuntimeConfig._1.jupyterUserScriptUri,
+              Set.empty, //TODO: remove instances from ListResponse
```

Comment:
Is this also to avoid API breaking changes?

Reply:
yea. Previously listResponse is using `unmarshallMinimalCluster`, which had all these fields stubbed out in db response.